dgisim/tests/test_dices.py

```diff
@@ -167,101 +168,438 @@ def test_just_satisfy_typical_all_omni(self):
+                    dices={Element.ANY: number_of_any}
+                )
+                input_precedence: list[set[Element]] = [
+                    {el} for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+                ]
+
+                actual: Optional[ActualDices] = input_actual_dices.smart_selection(
+                    requirement=input_abstract_dices, game_state=None, local_precedence_arg=input_precedence
+                )
+
+                if Dices._NUMBER_OF_ELEMENTS >= number_of_any:
+                    expected: Optional[ActualDices] = ActualDices(
+                        {el: 1 for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY[-number_of_any:]}
+                    )
+                else:
+                    expected = None
+
+                self.assertEqual(actual, expected)
+
+    def test_smart_selection_all_any_plus_one_omni_actual(self):
+        """test one die of every element from a to b, b-a+1 ANY
+        a<b"""
+        elements_and_omni = [Element.OMNI] + list(Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY)
+        for a in range(len(elements_and_omni)):
+            for b in range(a + 1, len(elements_and_omni)):
+                # need = supply, need < supply, need > supply
+                for any_need_number in [(b - a + 1), (b - a), (b - a + 2)]:
+                    with self.subTest(a=a, b=b, any_need_number=any_need_number):
+                        input_actual_dices: ActualDices = ActualDices(
+                            dices={el: 1 for el in elements_and_omni[a:b + 1]}
+                        )
+                        input_abstract_dices: AbstractDices = AbstractDices(
+                            dices={Element.ANY: any_need_number}
+                        )
+                        input_precedence: list[set[Element]] = [
+                            {el} for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+                        ]
+
+                        actual: Optional[ActualDices] = input_actual_dices.smart_selection(
+                            requirement=input_abstract_dices, game_state=None,
+                            local_precedence_arg=input_precedence
+                        )
+
+                        if any_need_number == (b - a + 1):  # =
+                            expected: Optional[ActualDices] = input_actual_dices
+                        elif any_need_number == (b - a):  # need < supply
+                            expected = ActualDices(
+                                dices={el: 1 for el in elements_and_omni[a + 1:b + 1]})
+                        elif any_need_number == (b - a + 2):
+                            expected = None
+                        self.assertEqual(actual, expected)
+
+    def test_smart_selection_concrete_element(
+            self, element: Element = Element.GEO):
+        """test fill element with itself
+        needed=1..10
+        supply=1..10"""
```

Comment:
again, for multi-line documentation, i usually do like
```py
"""
some words
more comments
"""
```

Reply:
done